### ISSUE TITLE
About accuracy

### ISSUE THREAD

**Issue by yijie422** [megvii-research/DPGN#9]
Thanks so much for your extraordinary work! I get a very low accuracy using your code. And I only changed num_workers from 8 to 0. 

This is my log.txt in 5way_1shot_resnet12_mini-imagenet:
[2020-07-08 00:57:47,863] [main] test_acc : 0.5358599852919579         step : 48000 
[2020-07-08 00:57:47,864] [main] test_best_acc : 0.5789199849367141    step : 13000
......
[2020-07-08 01:09:22,572] [main] -------------command line arguments-------------
[2020-07-08 01:09:22,572] [main] Namespace(checkpoint_dir='./checkpoints', config='config/5way_1shot_resnet12_mini-imagenet.py', dataset_root='dataset', device='cuda:0', display_step=100, exp_name='5way_1shot_resnet12_mini-imagenet', log_dir='./logs', log_step=100, mode='eval', num_gpu=1, seed=222)
[2020-07-08 01:09:22,572] [main] -------------configs-------------
[2020-07-08 01:09:22,572] [main] OrderedDict([('dataset_name', 'mini-imagenet'), ('num_generation', 6), ('num_loss_generation', 3), ('generation_weight', 0.2), ('point_distance_metric', 'l2'), ('distribution_distance_metric', 'l2'), ('emb_size', 128), ('backbone', 'resnet12'), ('train_config', OrderedDict([('num_ways', 5), ('num_shots', 1), ('batch_size', 25), ('iteration', 100000), ('lr', 0.001), ('weight_decay', 1e-05), ('dec_lr', 17000), ('dropout', 0.1), ('loss_indicator', [1, 1, 0]), ('lr_adj_base', 0.1), ('num_queries', 1)])), ('eval_config', OrderedDict([('num_ways', 5), ('num_shots', 1), ('batch_size', 10), ('iteration', 1000), ('interval', 1000), ('num_queries', 1)]))])
[2020-07-08 01:09:22,793] [main] find a checkpoint, loading checkpoint from ./checkpoints/5way_1shot_resnet12_mini-imagenet
[2020-07-08 01:09:27,521] [main] best model pack loaded
[2020-07-08 01:09:27,548] [main] current best test accuracy is: 0.5789199849367141, at step: 13000
[2020-07-08 01:10:52,549] [main] ------------------------------------
[2020-07-08 01:10:52,551] [main] step : 13000  test_edge_loss : 2.522496324658394  test_node_acc : 0.5824999854266644
[2020-07-08 01:10:52,551] [main] evaluation: total_count=999, accuracy: mean=58.25%, std=8.37%, ci95=0.52%

And in 5way_5shot_resnet12_mini-imagenet:
[2020-07-07 13:46:28,276] [main] -------------configs-------------
[2020-07-07 13:46:28,277] [main] OrderedDict([('dataset_name', 'mini-imagenet'), ('backbone', 'resnet12'), ('emb_size', 128), ('num_generation', 6), ('num_loss_generation', 6), ('generation_weight', 0.2), ('point_distance_metric', 'l2'), ('distribution_distance_metric', 'l2'), ('train_config', OrderedDict([('num_ways', 5), ('num_shots', 5), ('batch_size', 8), ('iteration', 100000), ('lr', 0.001), ('weight_decay', 1e-05), ('dec_lr', 15000), ('dropout', 0.1), ('lr_adj_base', 0.1), ('loss_indicator', [1, 1, 1]), ('num_queries', 1)])), ('eval_config', OrderedDict([('num_ways', 5), ('num_shots', 5), ('batch_size', 4), ('iteration', 1000), ('interval', 1000), ('num_queries', 1)]))])
[2020-07-07 13:46:28,445] [main] find a checkpoint, loading checkpoint from ./checkpoints/5way_5shot_resnet12_mini-imagenet
[2020-07-07 13:46:34,063] [main] best model pack loaded
[2020-07-07 13:46:34,090] [main] current best test accuracy is: 0.7321500130593777, at step: 17000
[2020-07-07 13:48:42,438] [main] ------------------------------------
[2020-07-07 13:48:42,439] [main] step : 17000  test_edge_loss : 4.024368638753891  test_node_acc : 0.7245500129163265
[2020-07-07 13:48:42,440] [main] evaluation: total_count=999, accuracy: mean=72.46%, std=12.81%, ci95=0.79%

How should I do to get a higher accuracy?

**Comment by zilunzhang** [comment]
Hi Yijie,

We did reproduce the problem you encountered, but we are still digging what mechanism causes this problem. Since the difference between num_worker = 0 and num_worker > 0 is creating [multi-worker processes](https://pytorch.org/docs/stable/data.html#single-and-multi-process-data-loading) or not, we deduce this problem may be a PyTorch bug.

If you set num_worker = 0 only because of the RAM limitation, we would suggest you set num_worker = 1, which will use the same amount of ram space comparing with num_worker = 0 (but accelerate the training) and keep the accuracy over 67% (tested).

Thanks,

DPGN team.

**Comment by zilunzhang** [comment]
> Thanks so much for your extraordinary work! I get a very low accuracy using your code. And I only changed num_workers from 8 to 0.
> 
> This is my log.txt in 5way_1shot_resnet12_mini-imagenet:
> [2020-07-08 00:57:47,863] [main] test_acc : 0.5358599852919579 step : 48000
> [2020-07-08 00:57:47,864] [main] test_best_acc : 0.5789199849367141 step : 13000
> ......
> [2020-07-08 01:09:22,572] [main] -------------command line arguments-------------
> [2020-07-08 01:09:22,572] [main] Namespace(checkpoint_dir='./checkpoints', config='config/5way_1shot_resnet12_mini-imagenet.py', dataset_root='dataset', device='cuda:0', display_step=100, exp_name='5way_1shot_resnet12_mini-imagenet', log_dir='./logs', log_step=100, mode='eval', num_gpu=1, seed=222)
> [2020-07-08 01:09:22,572] [main] -------------configs-------------
> [2020-07-08 01:09:22,572] [main] OrderedDict([('dataset_name', 'mini-imagenet'), ('num_generation', 6), ('num_loss_generation', 3), ('generation_weight', 0.2), ('point_distance_metric', 'l2'), ('distribution_distance_metric', 'l2'), ('emb_size', 128), ('backbone', 'resnet12'), ('train_config', OrderedDict([('num_ways', 5), ('num_shots', 1), ('batch_size', 25), ('iteration', 100000), ('lr', 0.001), ('weight_decay', 1e-05), ('dec_lr', 17000), ('dropout', 0.1), ('loss_indicator', [1, 1, 0]), ('lr_adj_base', 0.1), ('num_queries', 1)])), ('eval_config', OrderedDict([('num_ways', 5), ('num_shots', 1), ('batch_size', 10), ('iteration', 1000), ('interval', 1000), ('num_queries', 1)]))])
> [2020-07-08 01:09:22,793] [main] find a checkpoint, loading checkpoint from ./checkpoints/5way_1shot_resnet12_mini-imagenet
> [2020-07-08 01:09:27,521] [main] best model pack loaded
> [2020-07-08 01:09:27,548] [main] current best test accuracy is: 0.5789199849367141, at step: 13000
> [2020-07-08 01:10:52,549] [main] ------------------------------------
> [2020-07-08 01:10:52,551] [main] step : 13000 test_edge_loss : 2.522496324658394 test_node_acc : 0.5824999854266644
> [2020-07-08 01:10:52,551] [main] evaluation: total_count=999, accuracy: mean=58.25%, std=8.37%, ci95=0.52%
> 
> And in 5way_5shot_resnet12_mini-imagenet:
> [2020-07-07 13:46:28,276] [main] -------------configs-------------
> [2020-07-07 13:46:28,277] [main] OrderedDict([('dataset_name', 'mini-imagenet'), ('backbone', 'resnet12'), ('emb_size', 128), ('num_generation', 6), ('num_loss_generation', 6), ('generation_weight', 0.2), ('point_distance_metric', 'l2'), ('distribution_distance_metric', 'l2'), ('train_config', OrderedDict([('num_ways', 5), ('num_shots', 5), ('batch_size', 8), ('iteration', 100000), ('lr', 0.001), ('weight_decay', 1e-05), ('dec_lr', 15000), ('dropout', 0.1), ('lr_adj_base', 0.1), ('loss_indicator', [1, 1, 1]), ('num_queries', 1)])), ('eval_config', OrderedDict([('num_ways', 5), ('num_shots', 5), ('batch_size', 4), ('iteration', 1000), ('interval', 1000), ('num_queries', 1)]))])
> [2020-07-07 13:46:28,445] [main] find a checkpoint, loading checkpoint from ./checkpoints/5way_5shot_resnet12_mini-imagenet
> [2020-07-07 13:46:34,063] [main] best model pack loaded
> [2020-07-07 13:46:34,090] [main] current best test accuracy is: 0.7321500130593777, at step: 17000
> [2020-07-07 13:48:42,438] [main] ------------------------------------
> [2020-07-07 13:48:42,439] [main] step : 17000 test_edge_loss : 4.024368638753891 test_node_acc : 0.7245500129163265
> [2020-07-07 13:48:42,440] [main] evaluation: total_count=999, accuracy: mean=72.46%, std=12.81%, ci95=0.79%
> 
> How should I do to get a higher accuracy?

Hi Yijie,

We did reproduce the problem you encountered, but we are still digging what mechanism causes this problem. Since the difference between num_worker = 0 and num_worker > 0 is creating multi-worker processes or not, we deduce this problem may be a PyTorch bug.

If you set num_worker = 0 only because of the RAM limitation, we would suggest you set num_worker = 1, which will use the same amount of ram space comparing with num_worker = 0 (but accelerate the training) and keep the accuracy over 67% (tested).

Thanks,

DPGN team.

**Comment by yijie422** [comment]
Thanks for your reply. I also get 67% accuracy to set num_worker = 1. It's may be really a Pytorch bug.
